### PR TITLE
Levenshtein & Hamming Distance

### PR DESCRIPTION
**Pull request by c5sire**
This adds the Levenshtein function. It uses an algorithm that does not use an internal array and should be faster and more memory efficient. For the time being, string length is limited to 254. 
Addresses issue #1435.

Also, a simple mismatch counting function was added addressing issue #1439.

### CODE REVIEW

**Comment by c5sire** [comment]
I am trying to use the automated format fix script, but apparently it is not working on my fork. Any advice?
Best,
Reinhard

? make format-fix
python3 scripts/format.py --all --fix
Formatting all files
The files listed above will be reformatted.
Continue with changes (y/n)?
y
sh: 1: cmake-format: not found
Traceback (most recent call last):
  File "scripts/format.py", line 270, in <module>
    format_directory('src')
  File "scripts/format.py", line 265, in format_directory
    f.split('.')[-1])
  File "scripts/format.py", line 227, in format_file
    new_text = get_formatted_text(f, full_path, directory, ext)
  File "scripts/format.py", line 210, in get_formatted_text
    proc = subprocess.Popen(proc_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
  File "/mnt/d/apps/ubuntu/bin/miniconda3/lib/python3.7/subprocess.py", line 800, in __init__
    restore_signals, start_new_session)
  File "/mnt/d/apps/ubuntu/bin/miniconda3/lib/python3.7/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'cmake-format': 'cmake-format'
make: *** [Makefile:157: format-fix] Error 1
? make format
make: Nothing to be done for 'format'.

**Comment by Mytherin** [comment]
You need to install cmake-format, try `pip install cmake-format`. We should probably skip formatting cmake files if cmake-format is not installed instead of failing.

File: src/function/scalar/string/levenshtein.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        return -1; // too short
+    }
+	if (txt.size() > 254 || tgt.size() > 254) 
+    {
+        return -2; // too long
+    }
+
+    std::vector<int> distances(txt.size() + 1, 0);
+
+    for (int pos = 1; pos <= (int)txt.size(); pos++)
```
Comment (Mytherin) [inline src/function/scalar/string/levenshtein.cpp:26]: Could you change the indexes used in the different loops to `idx_t` rather than `int`? We prefer to avoid `int` since its size is platform-dependent.

File: src/function/scalar/string/levenshtein.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+// See: https://www.kdnuggets.com/2020/10/optimizing-levenshtein-distance-measuring-text-similarity.html
+// And: Iterative 2-row algorithm: https://en.wikipedia.org/wiki/Levenshtein_distance
+// Note: A first implementation using the array algorithm version resulted in an error raised by duckdb (too muach memory usage) 
+static int levenshtein2(const string& txt, const string& tgt) 
+    {
+    if (txt.size() < 1 || tgt.size() < 1) 
+    {
+        return -1; // too short
+    }
+	if (txt.size() > 254 || tgt.size() > 254)
```
Comment (Mytherin) [inline src/function/scalar/string/levenshtein.cpp:19]: What exactly is the reason for the 255 byte limitation? It seems like the code below works fine for larger strings, no?

File: src/function/scalar/string/levenshtein.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        distances[pos] = std::min(distances[pos - 1], pos - 1);
+        if (txt[pos - 1] != tgt[0]) 
+        {
+            distances[pos] += 1;
+        } 
+    }
+
+    int dist = 0;
+	int tempDist = 0;
```
Comment (Mytherin) [inline src/function/scalar/string/levenshtein.cpp:36]: idx_t here as well please :)

File: src/function/scalar/string/levenshtein.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+#include <ctype.h>
+#include <algorithm>
+
+namespace duckdb {
+
+// See: https://www.kdnuggets.com/2020/10/optimizing-levenshtein-distance-measuring-text-similarity.html
+// And: Iterative 2-row algorithm: https://en.wikipedia.org/wiki/Levenshtein_distance
+// Note: A first implementation using the array algorithm version resulted in an error raised by duckdb (too muach memory usage) 
+static int levenshtein2(const string& txt, const string& tgt)
```
Comment (Mytherin) [inline src/function/scalar/string/levenshtein.cpp:13]: We prefer to avoid the use of `std::string` since it performs a large number of (unnecessary) allocations. Could you change this function to operate on our internal `string_t` type instead? The resulting code would look very similar, only using `GetSize()` instead of `size()`, and `GetDataUnsafe()` to obtain a raw pointer to the data (note that string_t is not null-terminated).

File: src/function/scalar/string/mismatches.cpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+		return -1;  // Expected error code: both strings must be of equal length
+	}
+	if (num_characters_str < 1) {
+		return -1; // too shorts
+	}
+
+	int64_t mismatches = 0;
+
+	for (int64_t idx = 0; idx < (int64_t)str.GetSize(); ++idx) {
+		if (str.GetString()[idx] != tgt.GetString()[idx]) {
```
Comment (Mytherin) [inline src/function/scalar/string/mismatches.cpp:22]: Please avoid using `GetString()`, especially in a loop such as this: two new strings will be constructed for every character in the string. Instead, use `GetDataUnsafe()` and store the resulting pointer in a temporary variable, e.g.:


```cpp
auto str_data = str.GetDataUnsafe();
auto tgt_data = tgt.GetDataUnsafe();
for (...) {
    if (str_data[idx] != tgt_data[idx]) { ... }
}
```

File: src/function/scalar/string/mismatches.cpp [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+#include <ctype.h>
+#include <algorithm>
+
+namespace duckdb {
+
+static int64_t MismatchesScalarFunction(Vector &result, const string_t str, string_t tgt) {
+	int64_t num_characters_str = LengthFun::Length<string_t, int64_t>(str);
+	int64_t num_characters_tgt = LengthFun::Length<string_t, int64_t>(tgt);
+	if (num_characters_str != num_characters_tgt) {
+		return -1;  // Expected error code: both strings must be of equal length
```
Comment (Mytherin) [inline src/function/scalar/string/mismatches.cpp:13]: If you wish to throw an error from a function you should throw an exception, e.g. `throw InvalidInputException("...")`. This will report the error to the user.

File: test/sql/function/string/test_levenshtein.test [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+statement ok
+CREATE TABLE strings(s VARCHAR)
+
+statement ok
+INSERT INTO strings VALUES ('hello'), ('hallo'), ('aloha'), ('hoi'), ('world')
+
+# normal length
+query I
+SELECT levenshtein(s, 'hallo') FROM strings ORDER BY s
```
Comment (Mytherin) [inline test/sql/function/string/test_levenshtein.test:16]: I would like to see a number of extra tests, including `NULL` values, and each of the different combinations of (CONSTANT, CONSTANT), (COLUMN, CONSTANT), (CONSTANT, COLUMN), (COLUMN, COLUMN).

In addition, I would like to see tests that trigger each of the possible error conditions (too short strings, empty strings, too long strings).

File: test/sql/function/string/test_mismatches.test [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+statement ok
+CREATE TABLE strings(s VARCHAR)
+
+statement ok
+INSERT INTO strings VALUES ('hello'), ('hallo'), ('aloha'), ('hoi'), ('world'), (NULL)
+
+# normal length
+query I
+SELECT mismatches(s, 'hallo') FROM strings ORDER BY s
```
Comment (Mytherin) [inline test/sql/function/string/test_mismatches.test:16]: Same here, please add more tests as described above

**Review by Mytherin — COMMENTED** [review]
Thanks for your PR! This is looking very good! I have the following comments on the PR. When you have time could you have a look at them? Cheers.

File: src/function/scalar/string/levenshtein.cpp (hunk 1 above)
Comment (c5sire) [reply src/function/scalar/string/levenshtein.cpp:26]: Certainly.

File: src/function/scalar/string/levenshtein.cpp (hunk 2 above)
Comment (c5sire) [reply src/function/scalar/string/levenshtein.cpp:19]: No specific reason. Just thought it might be cautious. I will remove it.

File: src/function/scalar/string/levenshtein.cpp (hunk 4 above)
Comment (c5sire) [reply src/function/scalar/string/levenshtein.cpp:13]: Thanks for the pointers.

File: src/function/scalar/string/mismatches.cpp (hunk 5 above)
Comment (c5sire) [reply src/function/scalar/string/mismatches.cpp:22]: Thanks!

File: src/function/scalar/string/mismatches.cpp (hunk 6 above)
Comment (c5sire) [reply src/function/scalar/string/mismatches.cpp:13]: ok

File: test/sql/function/string/test_levenshtein.test (hunk 7 above)
Comment (c5sire) [reply test/sql/function/string/test_levenshtein.test:16]: Noted.

**Comment by c5sire** [comment]
Addressed your review comments.
I noticed that after fixing formatting and tidying issues, there are now apparently inconsistent failures with some environments like windows32, windows64, and Raspberry Pi. I didn't see any direct connection with the code I was working on, though.

**Comment by hannesmuehleisen** [comment]
Indeed the only failing test that remains is the Windows 64 bit test. I see
````
D:/a/duckdb/duckdb/test/api/test_progress_bar.cpp(36): FAILED:
117
  REQUIRE( con.context->progress_bar->IsPercentageValid() )
118
with expansion:
119
````

Maybe @pdet can have a look?

**Comment by c5sire** [comment]
Similarly a test failed for PR#1451 for windows32:
D:/a/duckdb/duckdb/test/api/test_progress_bar.cpp(36): FAILED:
  REQUIRE( con.context->progress_bar->IsPercentageValid() )
with expansion:
  false

**Comment by pdet** [comment]
Hi, yes, I can confirm that the progress bar is non-deterministically breaking on windows, I'm looking into it!

File: .clang-format [hunk 9]
```diff
@@ -15,7 +15,6 @@ SpaceBeforeCtorInitializerColon: true
 SpaceBeforeInheritanceColon: true
 SpacesInAngles: false
 SpacesInCStyleCastParentheses: false
-SpacesInConditionalStatement: false
```
Comment (Mytherin) [inline .clang-format:18]: Was this included by accident, or did you intend to change the formatting rules here? If so, could you expand on why?

File: src/function/scalar/string/levenshtein.cpp [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+#include "duckdb/common/vector_operations/vector_operations.hpp"
+#include "duckdb/common/string_util.hpp"
+
+#include <ctype.h>
+#include <algorithm>
+
+namespace duckdb {
+
+static idx_t Minimum(const idx_t &x, const idx_t &y, const idx_t &z) {
+	return std::min(x, std::min(y, z));
```
Comment (Mytherin) [inline src/function/scalar/string/levenshtein.cpp:11]: Please avoid using generic names like `Minimum` for functions. In DuckDB we support amalgamation builds which means all source files will be combined into a single `.cpp` file. Generic names are a potentially source of conflicts, since `Minimum` might be used somewhere else in the codebase. 

Since the function is only used in one place, I recommend inlining the function.

Also please use `MinValue` instead of std::min (from [duckdb/common/helper.hpp](https://github.com/cwida/duckdb/blob/master/src/include/duckdb/common/helper.hpp)). `std::min` has bitten us before on certain platforms.

**Review by Mytherin — COMMENTED** [review]
Thanks for the changes again, this is looking much better now. Some minor comments, and then I think this is ready to be merged:

File: .clang-format (hunk 9 above)
Comment (c5sire) [reply .clang-format:18]: It caused an error on my local installation and the script would not proceed.

File: src/function/scalar/string/levenshtein.cpp (hunk 10 above)
Comment (c5sire) [reply src/function/scalar/string/levenshtein.cpp:11]: Noted.

File: .clang-format (hunk 9 above)
Comment (Mytherin) [reply .clang-format:18]: Perhaps you are using an older version of clang-format? We are using version 11 (`clang-format version 11.0.1`).

Comment (c5sire) [reply .clang-format]: Re-added the line.

File: src/function/scalar/string/levenshtein.cpp (hunk 10 above)
Comment (c5sire) [reply src/function/scalar/string/levenshtein.cpp]: Used MinValue and simplified.

**Comment by Mytherin** [comment]
Thanks for all the changes, looks great now!

**Comment by c5sire** [comment]
Thanks to you!

**Comment by c5sire** [comment]
Hi, 
I noticed that my levenshtein, hamming, and jaccard functions do not work in the WHERE clause of a statement. While other text functions like length do. What did I miss? 
Best
